presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -138,6 +146,19 @@ public SqlQueryScheduler(QueryStateMachine queryStateMachine,
         this.executor = executor;
 
         rootStage.addStateChangeListener(state -> {
+            if (state != PLANNED && state != SCHEDULING) {
+                // We need all the exchange locations for the root stage to be available in one shot.
+                // If the stage is past the PLANNING and SCHEDULING state, all tasks are added and
+                // there won't be any new locations.
+                ImmutableSet.Builder<URI> uris = ImmutableSet.<URI>builder();
+                for (TaskInfo taskInfo : rootStage.getStageInfo().getTasks()) {
+                    List<BufferInfo> buffers = taskInfo.getOutputBuffers().getBuffers();
```

Comment:
IIRC there should only be one buffer per task for the output stage.  If so, we should enforce that here.